### PR TITLE
[dbnode] Sort by unmarked index entries for indexing

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This ensures that batches are ordered by whether entries are marked done or not to ensure proper batched metrics are indexed per block irregardless of errors in any single one batch.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/storage/index/types.go [hunk 1]
```diff
@@ -711,11 +711,11 @@ func (b *WriteBatch) Less(i, j int) bool {
 		panic(fmt.Errorf("unexpected sort by: %d", b.sortBy))
 	}
 
-	if b.entries[i].OnIndexSeries != nil && b.entries[j].OnIndexSeries == nil {
-		// This other entry has already been marked and this hasn't
+	if !b.entries[i].result.Done && b.entries[j].result.Done {
```
Comment (arnikola) [inline src/dbnode/storage/index/types.go:714]: 🎉

File: src/dbnode/storage/index.go [hunk 2]
```diff
@@ -320,6 +323,8 @@ func newNamespaceIndexWithOptions(
 
 		queryWorkersPool: newIndexOpts.opts.QueryIDsWorkerPool(),
 		metrics:          newNamespaceIndexMetrics(indexOpts, instrumentOpts),
+
+		debugSkipIndexEvery: newIndexOpts.opts.DebugSkipIndexEvery(),
```
Comment (notbdu) [inline src/dbnode/storage/index.go:327]: Maybe we should leave a TODO here to remove if we can figure out a better way to write the tests.

File: src/dbnode/storage/index.go [hunk 3]
```diff
@@ -1754,9 +1788,22 @@ func newNamespaceIndexMetrics(
 	opts index.Options,
 	iopts instrument.Options,
 ) nsIndexMetrics {
+	const (
+		indexAttemptName = "index-attempt"
+		forwardIndexName = "forward-index"
+	)
```
Comment (arnikola) [inline src/dbnode/storage/index.go:1794]: Does it matter if you define the const here or outside of the function?

File: src/dbnode/storage/index.go (hunk 2 above)
Comment (robskillington) [reply src/dbnode/storage/index.go]: Actually went and made a more useful thing which is to not index documents with certain field name/value pairs and using that now.

File: src/dbnode/storage/index.go (hunk 3 above)
Comment (robskillington) [reply src/dbnode/storage/index.go:1823]: It does not but this is nice that it's closer to the function in use, I really only do this with metric names.

**Pull request merged**